label_studio/io_storages/migrations/0017_auto_20240731_1638.py

```diff
@@ -1,40 +1,99 @@
+        "fk_constraint": "io_storages_redisexp_annotation_id_8547e508_fk_task_comp",
+        "column_name": "annotation_id"
+    },
+    {
+        "model_name": "s3exportstoragelink",
+        "table_name": "io_storages_s3exportstoragelink",
+        "index_name": "io_storages_s3exportstoragelink_annotation_id_729994fe",
+        "fk_constraint": "io_storages_s3export_annotation_id_729994fe_fk_task_comp",
+        "column_name": "annotation_id"
+    }
+]
 
 
 class Migration(migrations.Migration):
+    atomic = False
 
     dependencies = [
         ('tasks', '0047_merge_20240318_2210'),
         ('io_storages', '0016_add_aws_sse_kms_key'),
     ]
 
-    operations = [
-        migrations.AlterField(
-            model_name='azureblobexportstoragelink',
-            name='annotation',
-            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='io_storages_azureblobexportstoragelink', to='tasks.annotation'),
-        ),
-        migrations.AlterField(
-            model_name='gcsexportstoragelink',
-            name='annotation',
-            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='io_storages_gcsexportstoragelink', to='tasks.annotation'),
-        ),
-        migrations.AlterField(
-            model_name='localfilesexportstoragelink',
-            name='annotation',
-            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='io_storages_localfilesexportstoragelink', to='tasks.annotation'),
-        ),
-        migrations.AlterField(
-            model_name='redisexportstoragelink',
-            name='annotation',
-            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='io_storages_redisexportstoragelink', to='tasks.annotation'),
-        ),
-        migrations.AlterField(
-            model_name='s3exportstoragelink',
-            name='annotation',
-            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='io_storages_s3exportstoragelink', to='tasks.annotation'),
-        ),
-    ]
+    operations = []
+
+    for table in tables:
+        index_sql = create_index_sql(table['table_name'], table['index_name'], table['column_name'])
+        fk_sql = create_fk_sql(table['table_name'], table['fk_constraint'], table['column_name'], "task_completion", "id")
+
+        if IS_SQLITE:
+            operations.append(
+                migrations.AlterField(
+                    model_name=table['model_name'],
+                    name='annotation',
+                    field=models.ForeignKey(on_delete=models.deletion.CASCADE,
+                                            related_name=table['table_name'],
+                                            to='tasks.annotation'),
+                ),
+            )
+            continue
+
+        operations.append(
+            migrations.RunSQL(
+                sql=index_sql,
+                reverse_sql=RunSQL.noop
```

Comment:
Is it worth implementing the drop index / alter table drop constraint to enable a proper reverse migration?

Reply:
I added a `DROP INDEX` but I left the `noop` for the fk constraint since the reverse operation would involve re-adding a uniqueness constraint on the `annotation_id` column which may not be valid on the reversal. I did verify that the `IF NOT EXISTS ` and `IF EXISTS` keywords allow the forward migrations to work after migrations are reversed.

Reply:
Great point about re-adding the constraint no longer being valid, SGTM